Improve the upgrade documentation

joestringer opened the pull request:
Improve the upgrade documentation by:

* Adding an overview to describe what happens during upgrade, what the different files are, etc.
* Adding version-specific notes about what needs to occur to upgrade from Cilium version X to version Y.

Outstanding tasks, could be achieved in follow-up PR:
* [x] Update the 1.2 upgrade steps to document the potential issue with #5052 (limited prefix lengths on Linux < 4.11
* [x] Review `upgrade-impact` labelled issues for any additional known upgrade constraints:
  https://github.com/cilium/cilium/issues?utf8=%E2%9C%93&q=label%3Aupgrade-impact+

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5162)
<!-- Reviewable:end -->

danwent committed:
docs: Add overview of upgrade

Signed-off-by: Dan Wendlandt <dan@covalent.io>
Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
docs: Add outline for DaemonSet upgrade

Provide a brief overview of the approaches to upgrade and their
trade-offs, and separate the approaches into different subsections.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
docs: Create specific upgrade notes

Describe the process for upgrading between each supported Cilium
version.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
docs: Describe monitor aggregation downgrade impact

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm commented on Documentation/install/upgrade-1.1.rst:
```diff
@@ -0,0 +1,26 @@
+   * ``sidecar-istio-proxy-image``
+
+   See the :git-tree:`example ConfigMap
+   <examples/kubernetes/templates/v1/cilium-cm.yaml>` for more details.
```
These options no longer exist in the cilium-cm.yaml can the git-tree point to a specific branch?

joestringer replied:
AFAICT the above options exist both in the v1.1 branch and v1.2/master.

`:git-tree` already points to the "current" branch, eg if you're browsing 1.1 it will point to the example configmap for 1.1. It's defined based on `SCM_WEB`:

https://github.com/cilium/cilium/blob/693c576a89aafc11cad649266bf7f1be563637bc/Documentation/conf.py#L78-L86

joestringer replied:
Ah, there is `sidecar-http-proxy` which was deprecated in 1.1. I can mention that too.

aanm commented on Documentation/install/upgrade.rst:
```diff
@@ -209,35 +290,90 @@ Both files are dedicated to "\ |SCM_BRANCH|" for each Kubernetes version.
-command below, but be aware that copy of the spec file stored in Kubernetes
-might run out-of-sync with the CLI flags, or options, specified by each Cilium
-version.
+Direct version upgrade
```
I would add `(Not recommended for major or minor upgrades)`

joestringer replied:
I might put this in a note? Otherwise this bracketed part would be duplicated in each of the 1.0, 1.1, 1.2-specific instructions links.

aanm commented on Documentation/install/upgrade.rst:
```diff
@@ -209,35 +290,90 @@ Both files are dedicated to "\ |SCM_BRANCH|" for each Kubernetes version.
-    kubectl rollout status daemonset/cilium -n kube-system
+    $ kubectl rollout status daemonset/cilium -n kube-system
 
 To undo the rollout via rollback, run:
```
I think this should be in a different section, distracted users my run this blindly and they will undo the upgrade process in a single line without noticing.

joestringer replied:
I could put this in a nested "note" box or something so it's highlighted that this should only be necessary in unusual circumstances?

joestringer replied:
Actually, your original suggestion to put it in its own section would be better IMO.

aanm replied:
@joestringer There's already a Downgrade section below

joestringer replied:
Downgrade may occur during upgrade, or it could occur some time later (eg, someone realises a problem after 1 day). These instructions are here in case you need them during upgrade, whereas I've thought of the `Downgrade` section as somewhere that handles the more general case. That said, I think the downgrade section may need some additional love.

aanm reviewed (dismissed):
Maybe we should also add somewhere that it is recommended to upgrade to the latest patch version before upgrading to the next minor version?

joestringer committed:
docs: Split upgrade rollback into its own section

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
docs: Point upgrade specifics to latest versions

Provide a link to the latest version for a particular tree for each of
the 1.2, 1.1 and 1.0 upgrade sections.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
docs: Document downgrade for Cilium 1.1 and 1.0.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon commented on Documentation/install/upgrade.rst:
```diff
@@ -13,30 +13,96 @@ Upgrade Guide
+   resources by using kubectl to apply a new ``cilium-rbac.yaml`` associated
+   with your target version of Cilium.
+
+#. Update the DaemonSet resource by applying the ``cilium-ds.yaml`` associated
```
Add two backticks before and after DaemonSet

ianvernon commented on Documentation/install/upgrade.rst:
```diff
@@ -13,30 +13,96 @@ Upgrade Guide
+Upgrade ConfigMap (Recommended)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Upgrading the `ConfigMap` should be done manually before upgrading the ``RBAC``
```
Is use one backtick vs. two intentionally done for wrapping ConfigMap vs. other Kubernetes constructs in this section?

joestringer replied:
Somewhat intentional. One backtick links to the glossary section for it, two is just text literal. We don't have glossary items for all k8s resource types so I didn't consistently format them everywhere. I think a follow up could add those items and make this more consistent in general across the docs.

ianvernon reviewed (approved):
Nothing blocking. Thanks for doing this, the instructions are clear and well-written!
